File: common/src/main/kotlin/net/impleri/playerskills/skills/basic/BasicSkill.kt
package net.impleri.playerskills.skills.basic

import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.TeamMode
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation

open class BasicSkill(
  name: ResourceLocation,
  value: Boolean?,
  description: String?,
  options: List<Boolean>?,
  changesAllowed: Int?,
  teamMode: TeamMode?,
  notify: Boolean?,
  notifyKey: String?,
) : Skill<Boolean>(
  name,
  BasicSkillType.NAME,
  value,
  description,
  options ?: ArrayList(),
  changesAllowed ?: UNLIMITED_CHANGES,
  teamMode ?: TeamMode.off(),
  notify ?: false,
  notifyKey,
) {
  override fun copy(value: Boolean?, changesAllowed: Int): Skill<Boolean> {
    return BasicSkill(name, value, description, options, changesAllowed, teamMode, notify, notifyKey)
  }

  override fun getDefaultNotification(): Component {
    val messageKey =
      if (value == true) "playerskills.notify.basic_skill_enabled" else "playerskills.notify.basic_skill_disabled"
    val skillName = formatSkillName()
    return Component.translatable(messageKey, skillName)
  }
}

File: common/src/main/scala/net/impleri/playerskills/skills/basic/BasicSkill.scala
package net.impleri.playerskills.skills.basic

import net.impleri.playerskills.api.skills.ChangeableSkillOps
import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.TeamMode
import net.impleri.playerskills.facades.minecraft.core.ResourceLocation

case class BasicSkill(
  override val name: ResourceLocation,
  override val value: Option[Boolean] = None,
  override val description: Option[String] = None,
  override val options: List[Boolean] = List.empty,
  override val changesAllowed: Int = Skill.UNLIMITED_CHANGES,
  override val teamMode: TeamMode = TeamMode.Off(),
  override val announceChange: Boolean = false,
  override val notifyKey: Option[String] = None,
) extends Skill[Boolean] with ChangeableSkillOps[Boolean, BasicSkill] {
  override val skillType: ResourceLocation = BasicSkillType.NAME

  override protected[playerskills] def mutate(value: Option[Boolean], changesAllowed: Int): BasicSkill = {
    copy(value = value, changesAllowed = changesAllowed)
  }

  override def getMessageKey: String = {
    value match {
      case Some(true) => "playerskills.notify.basic_skill_enabled"
      case _ => "playerskills.notify.basic_skill_disabled"
    }
  }
}
